onedal/backend/data/data.cpp

```diff
@@ -0,0 +1,168 @@
+#define array_is_behaved_F(a) \
+    (PyArray_ISFARRAY_RO((PyArrayObject *)a) && array_type(a) < NPY_OBJECT)
+#define array_is_native(a) (PyArray_ISNOTSWAPPED((PyArrayObject *)a))
+#define array_numdims(a)   PyArray_NDIM((PyArrayObject *)a)
+#define array_data(a)      PyArray_DATA((PyArrayObject *)a)
+#define array_size(a, i)   PyArray_DIM((PyArrayObject *)a, i)
+
+#ifdef ONEDAL_DATA_PARALLEL
+int init_numpy() {
+    import_array();
+    return 0;
+}
+
+const static int numpy_initialized = init_numpy();
+#endif
+
+template <typename T>
+inline dal::homogen_table convert_from_numpy_to_homogen(PyArrayObject *array) {
+    std::int64_t column_count = 1;
+
+    if (array_numdims(array) > 2) {
+        throw std::runtime_error("Input array has wrong dimensionality (must be 2d).");
+    }
+    T *data_pointer = reinterpret_cast<T *>(array_data(array));
+    // TODO: check safe cast from int to std::int64_t
+    const std::int64_t row_count = static_cast<std::int64_t>(array_size(array, 0));
+    if (array_numdims(array) == 2) {
+        // TODO: check safe cast from int to std::int64_t
+        column_count = static_cast<std::int64_t>(array_size(array, 1));
+    }
+    const auto layout =
+        array_is_behaved_F(array) ? dal::data_layout::column_major : dal::data_layout::row_major;
+    auto res_table = dal::homogen_table(data_pointer,
+                                        row_count,
+                                        column_count,
+                                        dal::detail::empty_delete<const T>(),
+                                        layout);
+    // we need it increment the ref-count if we use the input array in-place
+    // if we copied/converted it we already own our own reference
+    if (reinterpret_cast<PyArrayObject *>(data_pointer) == array)
+        Py_INCREF(array);
+    return res_table;
+}
+
+dal::table convert_from_numpy_to_table(PyObject *obj) {
+    dal::table res;
+    if (obj == nullptr || obj == Py_None) {
+        return res;
+    }
+    if (is_array(obj)) {
+        PyArrayObject *ary = reinterpret_cast<PyArrayObject *>(obj);
+        if (array_is_behaved(ary) || array_is_behaved_F(ary)) {
+#define MAKE_HOMOGEN_TABLE(CType) res = convert_from_numpy_to_homogen<CType>(ary);
+            SET_NPY_FEATURE(PyArray_DESCR(ary)->type,
+                            MAKE_HOMOGEN_TABLE,
+                            throw std::invalid_argument("Found unsupported array type"));
+#undef MAKE_HOMOGEN_TABLE
+        }
+        else {
+            throw std::invalid_argument(
+                "[convert_from_numpy_to_table] Numpy input Could not convert Python object to onedal table.");
+        }
+    }
+    else {
+        throw std::invalid_argument(
+            "[convert_from_numpy_to_table] Not avalible input format for convert Python object to onedal table.");
+    }
+    return res;
+}
+
+void free_cap(PyObject *cap) {

```

Comment:
Not see the reason why the name is shortened. `free_capsule`? 
Also, the name does not reflect that it frees only special cases of capsules